src/python/pants/core/subsystems/lockfile_diff.py

```diff
@@ -0,0 +1,98 @@
+# Copyright 2022 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+from pants.backend.python.goals import lockfile
+from pants.backend.python.goals.lockfile import (
+    GeneratePythonLockfile,
+    GeneratePythonToolLockfileSentinel,
+)
+from pants.backend.python.subsystems.python_tool_base import PythonToolBase
+from pants.backend.python.target_types import ConsoleScript
+from pants.backend.python.util_rules import pex
+from pants.backend.python.util_rules.pex import PexRequest, VenvPex, VenvPexProcess
+from pants.core.goals.generate_lockfiles import (
+    GenerateLockfileResult,
+    GenerateToolLockfileSentinel,
+    LockfileGeneratedPostProcessing,
+)
+from pants.engine.process import InteractiveProcess, Process
+from pants.engine.rules import Get, collect_rules, rule
+from pants.engine.unions import UnionRule
+from pants.option.option_types import ArgsListOption, SkipOption
+from pants.util.docutil import git_url
+
+
+class LockfileDiffSubsystem(PythonToolBase):
+    options_scope = "lockfile-diff"
+    name = "lockfile-diff"
+    help = "The utility for diff-ing lockfiles."
```

Comment:
It seems very heavyweight to delegate this to a separate tool. Is the diffing logic so complex that it can't be run as inline code inside Pants?

Reply:
I agree, and the answer is two-fold I guess, as I've found use for this feature as a standalone tool as well, and as such it would be hard to get at embedded in Pants for those cases, but my initial idea was to have this as part of Pants itself (and now I've lost what the second case would've been..) but no, the diff'ing isn't super complex ([impl](https://github.com/kaos/lockfile-diff))

Perhaps we could find a hybrid solution where the diff'ing logic is available as a library which can be consumed (and provided?) by Pants as well as packaged as a cli tool?

Reply:
Implementing this standalone also allowed me to iterate faster on the implementation (just as an aside) as well as the initial version was based on in-house sources developed at iManage. (open sourced with permission)

Reply:
I'm still gonna nudge towards embedding this in Pants... :)

Reply:
OK, will do, but I think that will deserve a radically different implementation (but may still use similar approach) when using Pants constructs and APIs.

Reply:
I mean, I could be convinced to keep it separate if other maintainers agree...

Reply:
It's just that we believe in monorepos for a reason...

Reply:
Yea, I do see the benefit of having it in repo.. it was for this particular feature I see use cases beyond that of Pants, and in this way I can more easily re-use the common parts for both. But I'm open to take in more views on how we'd prefer this implemented :)

Reply:
If there is benefit to consuming the diff logic in 3rd party stuff, what if we publish another wheel (`python_distribution`) alongside pants? Then the code is in the pants monorepo, without relying heavily on pants internals. Then, a shim backend just imports the relevant module as a dependency, running it inline.

Anyway, just a drive by comment from the peanut gallery (with awful mixed metaphors to boot). Wdyt?

Reply:
yea, that's along the lines of what I had in mind with the comment from above:
> Perhaps we could find a hybrid solution where the diff'ing logic is available as a library which can be consumed (and provided?) by Pants

Reply:
Ended up with a pure Pants implementation. Not too much hassle to keep the `lockfile-diff` project alive separately if that turns out to fill a role (it does have a nice feature where I can diff dists from `PEX-INFO` files too, directly from within the `.pex` files! ;) )